docs/rfc/0698-cli-infrastructure.adoc

```diff
@@ -0,0 +1,297 @@
+and <<RFC8174>> when, and only when, they appear in all capitals, as shown here.
+
+
+== One `rad` to Rule Them All
+
+[quote]
+-------
+One ring to rule them all, one ring to find them,
+One ring to bring them all, and in the darkness bind them;
+-------
+-- J.R.R. Tolkein, Fellowship of the Ring
+
+The integral part of this proposal is that there will be one binary,
+`rad`, that will be the foundation for all other subcommands. It will
+define the global parameters and how to extend it.
+
+We will discuss these in the following two sections.
+
+== The Fellowship of the `rad`
+
+[quoute]
+--------
+“I can’t carry it for you, but I can carry you.” – Sam Gamgee
+--------
+-- J.R.R. Tolkien, The Return of the King
+
+To allow the ecosystem of commands to be extended by ourselves and
+others in the community it will be necessary to extend the `rad`
+command with further subcommands. The details of this are outlined in
+<<arch-cli, RFC-682#CLI>>, however, we will give a quick overview here for the sake
+of the reader.
+
+Taking inspiration from tools such as `git` and `cargo`, we propose
+that the `rad` binary can be extended by placing another
+binary on the users `$PATH` (or `%PATH%` on Windows). This
+binary MUST follow the naming convention of `rad-*`. To reiterate
+here, these new commands MUST NOT override the core set outlined here
+(nor in further RFCs).
+
+== Global Parameters
+
+Many subcommands will likely want to share a set of options. These
+options, dubbed as "global parameters" here, will be a defined set --
+exported from `radicle-link` for re-use -- that the base `rad` command
+will always receive and pass onto the subcommand being called.
+
+Other subcommands that are written in the target language (currently
+Rust) will be able to import the `struct` for reuse. If the subcommand
+is in another target language it SHALL add these parameters to its
+parameter list prefixed with `global-`. The `rad` command will parse
+these parameters, setting defaults if necessary, and SHALL pass them
+varbatim to the subcommand.
+
+Thus, all subcommands SHALL receive these global parameters as part of
+their interface. We list these parameters here, but it is NOT an
+exhaustive list:
+
+* `global-verbose`
+* `global-quiet`
+* `global-profile := active (default) | <profile-id>`
+
+== CLI Events
+
+The execution of a CLI subcommand could potentially result in an
+interesting event. For example, if Upstream is
+running and I create a project via the CLI, then a good user
+experience is that the Upstream is updated with the new project
+(without refreshing). In this case, the event is the creation of the
+project.
+
+In <<arch-pubsub, RFC-682#PubSub>>, a PubSub component was proposed --
+with a detailed specification being left to a future RFC. We will
+assume the existence of this component here. The CLI commands in this
+RFC MAY publish interesting events to the PubSub component. This will
```

Comment:
We're not defining the events here presumably. Are we planning on a future RFC listing all known events, or should there be a command which lists all the events a particular `rad` can produce? Put another way, if I'm a developer looking to listen to interesting events, how should I find out about them? What stability guarantees do I have?

Reply:
Good questions, I think at the least the events could be documented as part of the specific CLI library. But for stability guarantees, I think it'd be worth discussing in the future PubSub RFC, what do you reckon?

Reply:
Yeah I think this probably makes more sense in the PubSub RFC.